src/vmm/src/devices/virtio/net/test_utils.rs

```diff
@@ -244,17 +245,37 @@ pub fn enable(tap: &Tap) {
         .unwrap();
 }
 
+pub fn mock_frame(len: usize) -> Vec<u8> {
+    assert!(std::mem::size_of::<virtio_net_hdr_v1>() <= len);
+    let mut mock_frame = vec![b'A'; len];
+    // SAFETY:
+    // Frame is bigger than the header.
+    unsafe {
+        let hdr = &mut *mock_frame.as_mut_ptr().cast::<virtio_net_hdr_v1>();
+        let zeroed = std::mem::zeroed::<virtio_net_hdr_v1>();
+        *hdr = zeroed;
+        // We need to test num_buffers to 1 as the spec says.
+        hdr.num_buffers = 1;
+    }
+    mock_frame
```

Comment:
The way tests are set up is that they create random frame and then check that the content of the descriptors is same as the frame. Now our code properly set number of buffers, so we need to set it correctly in these random frames. Well, with this change they are not random anymore, because why were they random in the first place?

Reply:
There is no need to jump through all these hoops when `zerocopy` provides safe functions to do everything you want to do.

Reply:
Can you provide an example? Why `zerocopy` would have functions to allocate memory with certain alignment? Also since when memory allocation is a jump through a hoop?

Reply:
> Can you provide an example?

Something like

```rs
let mut hdr = virtio_net_hdr_v1::default();
hdr.num_buffers = 1;
let mut frame = vec![0; len];
frame[..size_of::<virtio_net_hdr_v1>()].copy_from_slice(hdr.as_bytes());
frame
```

you'd either need to implement `vm_memory::ByteValued` for `virtio_net_hdr_v1`, or derive `zerocopy::FromBytes` for it.

> Why zerocopy would have functions to allocate memory with certain alignment? 

It has safe functions for converting structs to byte arrays, which you can then memcpy into your Vec. No need to allocate anything at any alignment, because rustc will automatically get it right for you.

> Also since when memory allocation is a jump through a hoop?

Since here it required multiple unsafe operations when zero are needed

Reply:
If zero are needed, then we cannot use any of these as well, because they have `unsafe` inside them. I did update code here, so it is just 3 function calls. Do we really need to spend time choosing between 2 correct versions that do same thing?